gt/main.c

```diff
@@ -81,39 +155,107 @@ gt_proc(void *arg)
 			continue;
 
 		for (i = 0; i < num_rx; i++) {
+			int ret;
+			uint8_t priority;
+			struct rte_mbuf *m = rx_bufs[i];
+			struct gt_packet_fields fields;
+			struct ggu_policy policy;
+			struct ipip_tunnel_info tunnel;
+			struct ether_hdr *new_eth;
+
 			/*
-			 * TODO Decapsulate the packets.
+			 * Decapsulate the packets.
 			 *
 			 * Only request packets and priority packets
 			 * with capabilities about to expire go through a
 			 * policy decision.
 			 *
 			 * Other packets will be fowarded directly.
 			 */
+			ret = decapsulate(m, &priority, &tunnel);
+			if (ret < 0) {
+				rte_pktmbuf_free(m);
+				continue;
+			}
+
+			fields.outer_ip_ver = tunnel.flow.proto;
+			if (fields.outer_ip_ver == ETHER_TYPE_IPv4)
+				fields.gatekeeper_server_ip.v4 =
+					tunnel.flow.f.v4.src;
+			else if (fields.outer_ip_ver == ETHER_TYPE_IPv6)
+				rte_memcpy(fields.gatekeeper_server_ip.v6,
+					tunnel.flow.f.v6.src,
+					sizeof(fields.gatekeeper_server_ip.
+					v6));
```

Comment:
This block of code has a serious issue: `tunnel` is only updated in `decapsulate()` if priority is greater than 1, but this block of code always read `tunnel`. I think a better solution is to add a pointer to the the Ethernet header and another to outer IP header in `fields` to be updated by `decapsulate()`. This would push copying of fields to where it's needed.